### PR TITLE
[file] Fix relative symlinks to directories on Windows

### PR DESCRIPTION
**Pull request by staticfloat**
Windows has two concepts of directory symlinks: actual soft symlinks,
added in Visa, and directory junction points.  Junction points cannot
support relative symlinks, but that was what we were using if we
detected that the target was a directory.  This resulted in all sorts of
strange behavior, such as calling `symlink("foo", link)` creating a
junction point pointing to the absolute path `joinpath(pwd(), "foo")` if
there was a directory named `foo` in the current directory, and a
relative symlink to a file `"foo"` otherwise.

This PR cleans this up significantly, allowing for the creation of
realtive directory symlinks on Windows (assuming you have Administrator
privileges, of course) and even allowing users to force creation of a
directory symlink even if we can't auto-detect that the target should be
a directory.  My best advice is to create symlinks only after the files
themselves you want to point to already exist, but let's go ahead and
be flexible since it doesn't cost us much.

### CODE REVIEW

**Comment by staticfloat** [comment]
X-ref: https://github.com/JuliaLang/Pkg.jl/pull/2374
X-ref: https://github.com/JuliaLang/Pkg.jl/pull/2371

Many thanks to @IanButterworth for his help in tracking this down, and for performing the initial experiments when we thought we'd need to patch `libuv` to get this fixed.

**Comment by DilumAluthge** [comment]
Out of curiosity, will this patch fix the Windows CI problems on GHA?

**Comment by DilumAluthge** [comment]
Or are there other issues as well?

**Comment by staticfloat** [comment]
> Out of curiosity, will this patch fix the Windows CI problems on GHA?

It should at least address the problems I was running into at first; there may be more.  :)

**Comment by musm** [comment]
I find the optional keyword arg `dir_target` confusing. I think it would be better to remove this keyword argument. Instead I think we should simply *default* to symbolic links (these are more verstitile and more closely emulate unix symbolic links, so the platform differences are abstracted away in the default case) and instead add an optional  keyword argument  `junction=false` for those special use cases were someone explicitly needs to create  a junction point instead of a symbolic link.

**Comment by staticfloat** [comment]
`dir_target` doesn't switch between junctions and symbolic links; it allows a user to create a link (either a junction or a symbolic link) that points to a directory.  Without `dir_target=true`, if you try to create a symlink to a target that doesn't exist, it will create a file symlink, which results in all sorts of nastiness when you try to do anything with it once the target actually does exist.

**Comment by musm** [comment]
Understood. My point was that symbolic links can also point to directories, both absolute or relative.  So `dir_target` doesn't fully specify what kind of link should be created in the absolute path case, since a symbolic link would also be valid.

Changing to `junction=false` and doing additional error checking in the case `junction=true` is passed, would be a more clear user-facing option.

**Comment by staticfloat** [comment]
> Understood. My point was that symbolic links can also point to directories, both absolute or relative. So dir_target doesn't fully specify what kind of link should be created in the absolute path case, since a symbolic link would also be valid.

Even if we always used symlinks, we would _still_ need `dir_target` because you must create either a directory symlink or a file symlink; they are two completely different things on windows and are incompatible.

> Changing to junction=false and doing additional error checking in the case junction=true is passed, would be a more clear user-facing option.

As far as I can tell, using a symlink when you could instead use a junction is strictly worse; they are lower performance, especially over a network share, so if at all possible, we should always use junctions (which means whenever the target is an absolute path).

**Comment by musm** [comment]
Good point,  looks like that option would indeed be needed, my bad. 

> As far as I can tell, using a symlink when you could instead use a junction is strictly worse; they are lower performance, especially over a network share, so if at all possible, we should always use junctions (which means whenever the target is an absolute path).

My understanding is the opposite. Junction points are far more limited and are more "legacy". For example, they don't work across remote, relative, nor SMB files paths.  They "might" be more efficient, but worth to test. Symbolic links look to have been introduced on Vista. I guess the biggest annoyance with symlinks is that they require special privileges to create. In any case, Windows 10 has made this a lot easier where admin privileges are no longer required to create them. 

In any case, this later point is probably best reserved to a different PR?

**Comment by staticfloat** [comment]
> Junction points are far more limited and are more "legacy".

They are more limited, but given their advantages (able to be used on older versions of Windows, don't require Administrator privileges, higher performance) they should be used whenever possible.

**Comment by musm** [comment]
Symbolic links don't require admin privileges if you enable developer mode in Windows 10. 

Ref: https://blogs.windows.com/windowsdeveloper/2016/12/02/symlinks-windows-10/

Their support and performance have been enhanced. I think the claim for the higher performance of junction points needs to be checked/verified (it might not be significant). 


Microsoft dropped Windows Vista/7 support, as did Julia, so we can safely assume symlinks are available by default. 

Thus, I'm hesitant and it's not clear to me that junction points are indeed really the better default (again, junction points don't work across remote, relative, nor SMB files paths).  


I'm curious if @vtjnash  has anything to add to this discussion.

File: base/file.jl [hunk 1]
```diff
@@ -977,36 +977,77 @@ function sendfile(src::AbstractString, dst::AbstractString)
...
+            resolved_target = joinpath(dirname(link), target)
+        end
+
+        # If it is a directory (or `dir_target` is set), we'll need to add one
+        # of `UV_FS_SYMLINK_{DIR,JUNCTION}` to the flags, depending on whether
+        # `target` is an absolute path or not.
+        if (ispath(resolved_target) && isdir(resolved_target)) || dir_target
+            if isabspath(target)
+                flags |= UV_FS_SYMLINK_JUNCTION
+            else
```
Comment (musm) [inline base/file.jl:1039]: I guess an other option would be to `abspath(target)` if it is a relative path and then unconditionally call `flags |= UV_FS_SYMLINK_JUNCTION`

**Comment by staticfloat** [comment]
That’s essentially the old behavior and is what we don’t want; we want to
have the ability to have relative symlinks

**Comment by musm** [comment]
Yes, relative symlinks would still work. What I was hoping to get across, is to use `resolved_target` unconditionally.

```
    err = ccall(:jl_fs_symlink, Int32, (Cstring, Cstring, Cint), resolved_target, link, flags)

```
Is there  a problem with that?

**Comment by staticfloat** [comment]
> What I was hoping to get across, is to use resolved_target unconditionally.

`resolved_target` is not a relative path.  It's `joinpath(dirname(link), target)`.  So if `link` is `C:\dir\link` and `target` is `file`, then it will become `C:\dir\file` and it is an absolute path.  The reason we construct `resolved_target` is so that we can pass absolute paths to things like `stat()`, but we need to pass `target` (and only `target`) to `jl_fs_symlink` in order to get a relative symlink.

**Comment by IanButterworth** [comment]
That doctest fail looks odd. Could it be related?

**Comment by staticfloat** [comment]
I think it's https://github.com/JuliaLang/julia/commit/75b20b4441f6adc483d87bede57ae9b728930f7c

File: base/file.jl [hunk 2]
```diff
@@ -977,36 +977,77 @@ function sendfile(src::AbstractString, dst::AbstractString)
...
+        # it so that we can peek and see if it's a directory.
+        resolved_target = target
+        if !isabspath(target)
+            resolved_target = joinpath(dirname(link), target)
+        end
+
+        # If it is a directory (or `dir_target` is set), we'll need to add one
+        # of `UV_FS_SYMLINK_{DIR,JUNCTION}` to the flags, depending on whether
+        # `target` is an absolute path or not.
+        if (ispath(resolved_target) && isdir(resolved_target)) || dir_target
```
Comment (vtjnash) [inline base/file.jl:1036]:
```suggestion
        if dir_target || isdir(resolved_target)
```

File: test/file.jl [hunk 3]
```diff
@@ -29,23 +29,49 @@ end
...
+    nedlink = joinpath(subdir, "non_existant_dirlink")
+    symlink("new_dir", nedlink; dir_target=true)
+    try
+        readdir(nedlink)
+        @test false
+    catch e
+        @test isa(e, Base.IOError)
+        # It's surprisingly difficult to know what numeric value this will be across platforms
+        # so we'll just check the string representation instead. :(
+        @test endswith(e.msg, "(ENOENT)")
```
Comment (vtjnash) [inline test/file.jl:54]: Why not use the error constants in Base?

**Pull request merged**